### PR TITLE
Colorize spack info. Adds prominence to preferred version.

### LINKED ISSUE #2708: Preferred version and colorizing in `spack info`
`spack info` should show the preferred version in bold or color or something.  In general, some thoughtful colorizing of `spack info` o could be helpful.

### PR DESCRIPTION
**Pull request by alalazo**
fixes #2708

This uses `llnl.util.tty.color` to colorize the output of `spack info`. It also adds a section to display the preferred version before listing all the safe versions in order.

### CODE REVIEW

**Comment by adamjstewart** [comment]
Doesn't work for dark terminal themes:

screen shot 2017-08-07 at 10 39 24 am

I also find the version ordering very confusing.

> It also displays versions in the order the concretizer would choose them and shows the preferred in a line on its own and in bold.

It seems like 3.5.1 is on a line on its own, not the preferred version. I would prefer all versions in order (develop first, followed by numeric versions, followed by non-numeric versions). Maybe make the preferred version bold, or a different color?

**Comment by davydden** [comment]
> I would prefer all versions in order (develop first, followed by numeric versions, followed by non-numeric versions)

:+1:

> Doesn't work for dark terminal themes:

I also use dark themes

**Comment by adamjstewart** [comment]
> Maybe make the preferred version bold, or a different color?

Or even just an asterisk after the preferred/default version.

**Comment by alalazo** [comment]
@adamjstewart I'll try later removing the black colorization from the section titles, and making bold the preferred version (leaving the order lexicographical). Does it sound good?

**Comment by adamjstewart** [comment]
Yeah, give it a shot and I'll test it. Some terminals like the one on macOS allow you to deactivate support for bold fonts. It might be safer to put an asterisk before the preferred/default version (and make it bold?).

**Comment by alalazo** [comment]
I think this is ready for a second round. I can't really think of any test that can be added to satisfy codecov (and increase our coverage). If anyone has ideas, they are welcome.

**Comment by adamjstewart** [comment]
I'll take a look.

> I can't really think of any test that can be added to satisfy codecov (and increase our coverage).

It may sound stupid, but simply running `spack info` on a couple complicated packages would be better than nothing.

**Comment by alalazo** [comment]
@adamjstewart Make sense. I recently read a comment in a blog post that argued that python is a language where you need 100% coverage to be sure you don't have typos. As much as I like the language can't say it's not true. Of course if we could assert something it would be better...

**Comment by adamjstewart** [comment]
Color scheme looks much better on a dark terminal theme now!

We probably don't need to make the preferred version bold if it is a completely different section now.

This wasn't changed by your PR, but there are a couple things I never noticed that could be changed:
1. We definitely don't need a bunch of spaces around the homepage.
2. We probably don't need an empty line between each variant. This makes packages with a lot of variants have a very long info output (think `boost`).

**Comment by adamjstewart** [comment]
> I recently read a comment in a blog post that argued that python is a language where you need 100% coverage to be sure you don't have typos.

The number of times that I've seen a `raise Error(msg)` that was broken due to string formatting and was never discovered because most people don't get an error...

**Comment by davydden** [comment]
i just checked it and it looks good to me! 

> We probably don't need to make the preferred version bold if it is a completely different section now.

:+1:

> We definitely don't need a bunch of spaces around the homepage.

indeed, should be indented the same way as version's URLs.

> We probably don't need an empty line between each variant

agree here as well. 

While there, i would add that there appears to be two empty lines after variants and before Installation phase.

**Comment by davydden** [comment]
Matter of preference, but i would also keep preferred URL the same font colour/weight as other URLs as it's now in a separate section.

**Comment by alalazo** [comment]
I think the only one I don't agree with is:

> We probably don't need an empty line between each variant

To me this looks messier:
```console
    Name [Default]           Allowed values          Description

    cuda [off]               True, False             Enable CUDA support
    fabrics []               psm, psm2, pmi,         List of fabrics that are enabled
                             verbs, mxm              
    java [off]               True, False             Build Java support
    schedulers []            alps, lsf, tm,          List of schedulers for which
                             slurm, sge,             support is enabled
                             loadleveler             
    sqlite3 [off]            True, False             Build SQLite3 support
    thread_multiple [off]    True, False             Enable MPI_THREAD_MULTIPLE support
    vt [on]                  True, False             Build VampirTrace support

```
than this:
```console
    Name [Default]           Allowed values          Description


    cuda [off]               True, False             Enable CUDA support

    fabrics []               psm, psm2, pmi,         List of fabrics that are enabled
                             verbs, mxm              

    java [off]               True, False             Build Java support

    schedulers []            alps, lsf, tm,          List of schedulers for which
                             slurm, sge,             support is enabled
                             loadleveler             

    sqlite3 [off]            True, False             Build SQLite3 support

    thread_multiple [off]    True, False             Enable MPI_THREAD_MULTIPLE support

    vt [on]                  True, False             Build VampirTrace support
```
What do you think?

**Comment by adamjstewart** [comment]
Messier, slightly. Longer, definitely. I won't hold you to it either way, just a suggestion.

**Comment by davydden** [comment]
well, this looks too big to my taste:
```
Variants:
    Name [Default]               Allowed values          Description


    arpack [on]                  True, False             Compile with Arpack and PArpack
                                                         (only with MPI)

    build_type [DebugRelease]    Debug, Release,         The build type to build
                                 DebugRelease

    doc [off]                    True, False             Compile with documentation

    gsl [on]                     True, False             Compile with GSL

    hdf5 [on]                    True, False             Compile with HDF5 (only with MPI)

    int64 [off]                  True, False             Compile with 64 bit indices support

    metis [on]                   True, False             Compile with Metis

    mpi [on]                     True, False             Compile with MPI

    netcdf [on]                  True, False             Compile with Netcdf (only with MPI)

    oce [on]                     True, False             Compile with OCE

    optflags [off]               True, False             Compile using additional
                                                         optimization flags

    p4est [on]                   True, False             Compile with P4est (only with MPI)

    petsc [on]                   True, False             Compile with Petsc (only with MPI)

    python [on]                  True, False             Compile with Python bindings

    slepc [on]                   True, False             Compile with Slepc (only with Petsc
                                                         and MPI)

    trilinos [on]                True, False             Compile with Trilinos (only with
                                                         MPI)

```
I am ok either way, but i think it's ok to remove the line.

**Comment by davydden** [comment]
p.s. i think for `mpi` it looks messier due to multi-variants of `schedulers` and `schedulers`, whereas for `True/False` variants the eye catches the visual structure IMO much metter without empty lines.

**Comment by alalazo** [comment]
Ok, [complaint driven development](https://blog.codinghorror.com/complaint-driven-development/) for the win. I'll remove the line!

**Comment by adamjstewart** [comment]
Ran a quick:
```
$ for package in $(spack list); do echo $package; spack info $package > /dev/null; done
```
with this branch and the only package with an error message is a known issue with `hdf5-blosc`: #3676. I'll fix that in another PR.

File: lib/spack/spack/cmd/info.py [hunk 1]
```diff
@@ -48,11 +56,23 @@ def pad(string):
...
+    return spack.spec.version_color + s + plain_format
+
+
+def variant(s):
+    return spack.spec.enabled_variant_color + s + plain_format
 
 
 class VariantFormatter(object):
-    def __init__(self, variants, max_widths=(25, 20, 35)):
+    def __init__(self, variants, max_widths=(30, 20, 30)):
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/info.py]: If only there was a way to make this dynamic based on terminal width...

Comment (alalazo) [inline lib/spack/spack/cmd/info.py]: Let's leave it for another PR, but [here it is](https://stackoverflow.com/questions/566746/how-to-get-linux-console-window-width-in-python)

Comment (adamjstewart) [inline lib/spack/spack/cmd/info.py]: Completely agreed, I don't want to make this PR any more complicated.

File: lib/spack/spack/cmd/info.py [hunk 2]
```diff
@@ -186,8 +227,11 @@ def print_text_info(pkg):
                     inverse_map[when] = set()
                 inverse_map[when].add(spec)
         for when, specs in reversed(sorted(inverse_map.items())):
-            print("    %s provides %s" % (
-                when, ', '.join(str(s) for s in specs)))
+            line = "    %s provides %s" % (
+                when.colorized(), ', '.join(s.colorized() for s in specs)
```
Comment (alalazo) [inline lib/spack/spack/cmd/info.py]: Just a warning: I had a couple of weird python failures related to the colorization of the provided specs at the end of `spack info`. The failures happened only on `python@3.3` and `python@3.4`, and I couldn't reproduce them locally. Restarting the jobs apparently made Travis go green. Let me know if, for safety, you prefer to remove colorization or if we are fine like this. Other than that this is for me ready to be merged.

**Comment by alalazo** [comment]
@tgamblin @becker33 This changes the UI somehow, so I'll be waiting for your review to merge it.
